include/s3select_json_parser.h

```diff
@@ -0,0 +1,419 @@
+    Ch Peek() //const 
+    {
+      if(RAPIDJSON_UNLIKELY(src_ == end_))
+      {
+	if(next_src_)//next chunk exist
+	{//upon reaching to end of current buffer, to switch with next one
+	  src_ = next_src_;
+	  begin_ = src_;
+	  size_ =next_size_;
+	  end_ = src_ + size_;
+
+	  next_src_ = 0;
+	  next_size_ = 0;
+	  return *src_;
+	}
+	else return 0;
+      }
+      return *src_;
+    }
+
+    //override Take method
+    Ch Take() 
+    {
+      if(RAPIDJSON_UNLIKELY(src_ == end_))
+      {
+	if(next_src_)//next chunk exist
+	{//upon reaching to end of current buffer, to switch with next one
+	  src_ = next_src_;
+	  begin_ = src_;
+	  size_ = next_size_;
+	  end_ = src_ + size_;
+
+	  next_src_ = 0;
+	  next_size_ = 0;
+	  return *src_;
+	}
+	else return 0;
+      }
+      return *src_++;
+    }
+
+    void resetBuffer(char* buff, size_t size)
+    {
+      if(!src_)
+      {//first time calling
+	begin_ = buff;
+	src_ = buff;
+	size_ = size;
+	end_= src_ + size_;
+	return;
+      }
+
+      if(!next_src_)
+      {//save the next-chunk that will be used upon parser reaches end of current buffer
+	next_src_ = buff;
+	next_size_ = size;
+      }
+      else
+      {// should not happen
+	std::cout << "can not replace pointers!!!" << std::endl;//TODO exception
+	return;
+      }
+    }
+
+    void saveRemainingBytes()
+    {//this routine called per each new chunk
+      //savine the remaining bytes, before its overriden by the next-chunk.
+      size_t copy_left_sz = getBytesLeft(); //should be very small
+      internal_buffer.assign(src_,copy_left_sz);
+      
+      src_ = internal_buffer.data();
+      begin_ = src_;
+      size_ = copy_left_sz;
+      end_= src_ + copy_left_sz;
+    }
+
+    size_t getBytesLeft() { return end_ - src_; }
+
+};
+
+class JsonParserHandler : public rapidjson::BaseReaderHandler<rapidjson::UTF8<>, JsonParserHandler> {
+
+  public:
+
+    typedef enum {OBJECT_STATE,ARRAY_STATE} en_json_elm_state_t;
+
+    enum class row_state
+    {
+      ARRAY,
+      START_ROW,
+      OBJECT_START_ROW,
+      ARRAY_START_ROW,
+      ARRAY_END_ROW,
+      END_ROW
+    };
+
+    row_state state;
+
+    std::function <int(std::pair < std::string, Valuesax>)> fp;
+
+    std::vector <std::vector<std::string>> query_matrix{};
+    int row_count{};
+    std::vector <std::string> from_clause{};
+    bool prefix_match{};
+    Valuesax value;
+    ChunksStreamer stream_buffer;
+    bool init_buffer_stream;
+    rapidjson::Reader reader;
+    std::vector<en_json_elm_state_t> json_element_state;
+    std::string m_result;//debug purpose
+    std::vector<std::string> key_path;
+
+    JsonParserHandler() : init_buffer_stream(false)
+    {} 
+
+    std::string get_key_path()
+    {//for debug
+	  std::string res;
+	  for(const auto & i: key_path)
+	  {
+	    res.append(i);
+	    res.append(std::string("/"));
+	  }
+	  return res;
+    }
+
+    void dec_key_path()
+    {
+      if(json_element_state.back() != ARRAY_STATE)
```

Comment:
upon wrong input, it may cause `segfault`
the `json_element_state` size should be verified.